### PR TITLE
feat: add group allowlist for oidc

### LINKED ISSUE #10705: Add `CODER_OIDC_ALLOWED_GROUPS` to limit which groups can sign in to Coder
## Background

For GitHub, we have https://coder.com/docs/v2/latest/cli/server#--oauth2-github-allowed-orgs and https://coder.com/docs/v2/latest/cli/server#--oauth2-github-allowed-teams that only allows a subset of GitHub users to log in to Coder.

We do not have something similar for OIDC log ins.

## Proposal

- Add CODER_OIDC_ALLOWED_GROUPS which accepts a list of strings (group names).
- If a user is not in any of the groups, display an error, similar to the GitHub one
- Use https://coder.com/docs/v2/latest/cli/server#--oidc-group-field to define which groups claim is used to validate against
- Display a proper error/warning in the server if the claim is not being sent, similar to how we do for group sync

### PR DESCRIPTION
**Pull request by Emyrk**
Users not in the group allowlist cannot authenticate with Coder.

The authorized groups are never returned to the user, but I will return what groups the user is in to aid in debugging.

Closes https://github.com/coder/coder/issues/10705

# When you have 0 groups

Screenshot from 2023-12-06 15-59-43

# When you are in groups that are not authorized

Screenshot from 2023-12-06 16-17-02

### CODE REVIEW

**Comment by Emyrk** [comment]
Current dependencies on/for this PR:
* main
  * **PR #11070** <a href="https://app.graphite.dev/github/pr/coder/coder/11070?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/coder/coder/11070?utm_source=stack-comment).

File: coderd/userauth.go [hunk 1]
```diff
@@ -1009,14 +1010,23 @@ func (api *API) userOIDC(rw http.ResponseWriter, r *http.Request) {
...
 	// This is so we can support manual group assignment.
 	if api.OIDCConfig.GroupField != "" {
+		// allow list is a map of groups that are allowed to log in.
+		allowed := make(map[string]bool)
+		for _, group := range api.OIDCConfig.GroupAllowList {
+			allowed[group] = true
+		}
+		// If the allow list is empty, then the user is allowed to log in.
+		// Otherwise, they must belong to at least 1 group in the allow list.
+		inAllowList := len(allowed) == 0
```
[the remark this answers is not in the document]
Comment (Emyrk) [reply coderd/userauth.go:1028]: Yup it would be :+1:

**Review by johnstcn — APPROVED** [review]
This'll be handy for folks doing OIDC on a gmail domain for sure

**Comment by matifali** [comment]
@Emyrk can you also add some basic docs [here](https://coder.com/docs/v2/latest/admin/auth#openid-connect) on how to configure it in a follow up PR?

**Pull request merged**